### ISSUE TITLE
Hydra fails at runtime due to buildPerlPackage changes

### ISSUE THREAD

**Issue by lopsided98** [NixOS/nixpkgs#72783]
**Describe the bug**

Hydra fails at runtime due to missing transitive dependencies. This occurs because https://github.com/NixOS/nixpkgs/commit/9e60fefc349c9aa52927546e575889615b0f5ed5 removed `nix-support/propagated-user-env-packages` from all Perl packages. Hydra uses `buildEnv` to create an environment with all its Perl packages, and `propagated-user-env-packages` used to cause all the transitive dependencies to be pulled into the environment, but this no longer happens.

As suggested by @conferno, I tried replacing `buildEnv` with `perl.withPackages`, but this does not work because the Perl wrapper clobbers `PERL5LIB`, which [prevents Hydra from adding its own Perl files to `PERL5LIB`](https://github.com/NixOS/nixpkgs/blob/master/pkgs/development/tools/misc/hydra/default.nix#L122).

**To Reproduce**
```
nix-build nixos/tests/hydra
```

Maintainer information:
```yaml
# a list of nixpkgs attributes affected by the problem
attribute:
  - hydra
# a list of nixos modules affected by the problem
module:
  - hydra
```

cc @volth @Ma27

**Comment by conferno** [comment]
It needs also https://github.com/NixOS/nixpkgs/pull/59978 for `nix.perl-bindings` to work in `perl.withPackages`

**Comment by conferno** [comment]
Or just
```diff
--- a/pkgs/development/tools/misc/hydra/default.nix
+++ b/pkgs/development/tools/misc/hydra/default.nix
@@ -16,7 +16,8 @@ let                              
   perlDeps = buildEnv {                           
     name = "hydra-perl-deps";                     
     paths = with perlPackages;                    
-      [ ModulePluggable                           
+      (requiredPerlModules [                      
+        ModulePluggable                           
         CatalystActionREST                        
         CatalystAuthenticationStoreDBIxClass      
         CatalystDevel                             
@@ -65,7 +66,7 @@ let                              
         nix.perl-bindings                         
         git                                       
         boehmgc                                   
-      ];                                          
+      ]);                                         
   };                                              
 in stdenv.mkDerivation rec {                      
   pname = "hydra";                                
```
It will bring unlisted dependencies

**Comment by conferno** [comment]
`lib.closePropagation` instead of `requiredPerlModules` is even more harder, it will bring non-perl dependencies as well (such as `git`)

**Comment by Ma27** [comment]
The overall issue might be solved in the Perl infrastructure of nixpkgs, however I experienced exactly this when I prepared a Hydra bump last weekend (but lacked time to dig in). I'll prepare a PR to unbreak Hydra now, but don't have a strong opinion about how to fix this perl-wide.

**Comment by lopsided98** [comment]
[`rt`](https://github.com/NixOS/nixpkgs/blob/master/pkgs/servers/rt/default.nix) might have the same problem. I'm not sure if all transitive dependencies are explicitly included there.

**Comment by Ma27** [comment]
Thanks a lot for the insights @volth!

> The list of hydra's perl deps looks crafted manually including all transitive dependencies. As the time went on, the list got out of sync but nothing was broken, because buildEnv magically resolved perl dependencies, silently inserting missing ones. It might be nice, but the implicit magic is confusing. It resolves perl dependencies, but not python's, not ruby's, not lua'a

I actually tried to add all transitive perl dependencies of Hydra, but this escalated pretty quickly and I decided that this isn't a feasible approach since tracking the entire dependency graph makes the package IMHO rather hard to maintain.

> the code below should restore old behavior (just by adding a single word instead of ~10 lines on bash which also shadowed postFixup defined in buildPerlPackage), also being language-independent and giving an explicit hint of what is going on:

Agreed. This revert would cause a pretty big rebuild and I just don't think that's worth the effort.

IMHO it would be fine to add `lib.closePropagation` to automatically resolve all paths for now (to get Hydra fixed on master). Sooner or later someone (probably me as package-maintainer :sweat_smile: ) should take a closer look at the dependency-list to clean up outdated stuff.

As release isn't affected I'd also be fine to wait until the third solution works fine for us though.

**Comment by Ma27** [comment]
This should be fixed, at least with #83600 or probably even earlier. I'm currently running Hydra on 20.03 without any issues now. If you encounter further issues, feel free to ping me to reopen!
